### ISSUE TITLE
[Kotlin] EI_EXPOSE_REP not in line with Kotlin's concept for mutable/immutable collections?

### ISSUE THREAD

**Issue by Moritz5432** [spotbugs/spotbugs#3381]
### Issue Description

I have some doubts if the [EI_EXPOSE_REP](https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#ei-may-expose-internal-representation-by-returning-reference-to-mutable-object-ei-expose-rep) is really appropriate for Kotlin projects.

In my case it's complaining about DTO data classes which expose a kotlin “List” object like this:
```kotlin
@JsonIgnoreProperties(ignoreUnknown = true)
data class TransportParty(

    @JsonProperty("contacts")
    val contacts: List<Contact>
)
```

The rule is not fully wrong as a look into the byte code would reveal that a (mutable) ArrayList is actually used. However, Kotlin’s approach/design for immutability with lists is very different. 

The “List” interface makes sure to expose no mutating methods. If it is intended that the list should be mutable “MutableList” interface is used in Kotlin.

Additionally, Kotlin offers the “listOf” method to create “immutable” lists and “mutableListOf” method to create mutable lists. However, the actual implementation class may change with kotlin stdlib. Currently, it’s creating an ArrayList also for “listOf” with an interface which hides any mutating methods.

You can read more about it here: [The Mystery of Mutable Kotlin Collections | by Uberto Barbini | ProAndroidDev](https://proandroiddev.com/the-mystery-of-mutable-kotlin-collections-e82cbf5d781)

So even though, the rule (from byte code perspective) is not invalid it neglects the collection design which Kotlin follows and such issues are no real issues in a pure Kotlin project. 

### Environment

| Component    | Version |
| -------- | ------- |
| SonarQube  | 105476    |
| Sonar-FindBugs | 4.4.2     |
| Kotlin | 2.1.20    |

**Comment by gtoison** [comment]
Hello @Moritz5432 
I'm transferring this to the [SpotBugs](https://github.com/spotbugs/spotbugs) project because the underlying issue is flagged by SpotBugs.

It looks to me like Kotlin's `List` means "an immutable list", but at the same time we have `MutableList extends List`, so something declared as an "immutable list" might be a "mutable list" at runtime.
In other words, Kotlin's `List` semantic is: "mutations are not implemented", not: "mutations are blocked".

I don't know Kotlin enough to answer expertly, so please correct me if I'm wrong!

**Comment by Moritz5432** [comment]
Hi @gtoison ,
thanks for transferring it to the right place.

Your understanding is correct. And technically the rule is working correctly if we are talking only about the compiled classes.

However, it defies the general approach of Kotlin regarding immutability and creation of lists. Kotlin indicates via List / MutableList interfaces if a list is intended to be mutated or not. 
The recommended way to create an immutable list in Kotlin is to call `listOf()` (https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/list-of.html).
The recommended way to create a mutable list in Kotlin is to call `mutableListOf()` (https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/mutable-list-of.html)

It's up to Kotlin to decide what should be the exact List implementation used for such calls. As of now `listOf()` returns an `Arrays$ArrayList` but this could also change with future Kotlin versions.

I would argue here that the EI_EXPOSE_REP rule somehow "forces" us to go against the design decisions of Kotlin.

**Comment by gtoison** [comment]
Thanks for the clarifications @Moritz5432 
I think it would help a lot if you could provide a Kotlin sample where `EI_EXPOSE_REP` is expected (you have already provided one where it is NOT expected)

**Comment by Moritz5432** [comment]
Hi @gtoison 
sorry for the late reply. An example where it would be expected would be the following:
`@JsonIgnoreProperties(ignoreUnknown = true)
data class TransportParty(

    @JsonProperty("contacts")
    val contacts: MutableList<Contact>
)`
